server/proxy/httpproxy/director.go

```diff
@@ -33,20 +34,23 @@ func init() {
 	rand.Seed(time.Now().UnixNano())
 }
 
-func newDirector(lg *zap.Logger, urlsFunc GetProxyURLs, failureWait time.Duration, refreshInterval time.Duration, stop <-chan struct{}, donec chan<- struct{}) *director {
+func newDirector(lg *zap.Logger, urlsFunc GetProxyURLs, failureWait time.Duration, refreshInterval time.Duration) *director {
 	if lg == nil {
 		lg = zap.NewNop()
 	}
 	d := &director{
 		lg:          lg,
 		uf:          urlsFunc,
 		failureWait: failureWait,
+		stopc:       make(chan struct{}),
+		donec:       make(chan struct{}),
 	}
+	osutil.RegisterInterruptHandler(func() {
+		close(d.stopc)
```

Comment:
This at least tells the goroutine to stop. Should it also wait for it to finish (read from `donec`)?

Reply:
I think yes. We should wait, such as 1 second. 

It's better to get the handler included in a method, so that the unit test case can use as well.

Reply:
After second thought, I think it makes more sense to register the interrupt handler in higher level. It seems that it needs a little refactor, let me handle it. Please get this handler included in a method.

Reply:
No worries, thanks for thinking through this. I've since removed the interrupt handler registration.